# КОММЕНТАРИЙ РЕВЬЮВЕРА 2:

Приветствую тебя еще раз!) Спасибо за то, что доработал проект!

Мои комментарии будут также обозначены разными цветами, но с пометкой 2:
    
    «КОММЕНТАРИЙ РЕВЬЮВЕРА 2: ... »

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
Родион, ты хорошо выполнил доработку!) Думаю, теперь я могу принять работу, желаю успехов в дальнейшем обучении!)))


</div>

# КОММЕНТАРИЙ РЕВЬЮВЕРА
Хоу хоу хоу, Родион👋☃️🎄!)

Меня зовут Обидин Андрей, я осуществлял проверку проекта. Может перейдем сразу на <b>ты</b>? Если это неприемлемо, то обязательно дай знать)

Благодарю тебя за то, что вовремя всё сдал и хорошо поработал!) У тебя получился классный проект, однако некоторые моменты нужно слегка подправить, я оставил рекомендации, думаю у тебя все получистя!)

Далее по ходу работы ты сможешь встретить мои комментарии, они будут отмечены разными цветами:

<div class="alert alert-block alert-success">
Похвала и возможности развития
</div>
<div class="alert alert-block alert-warning">
Некритичные замечания, но стоит обратить внимание
</div>
<div class="alert alert-block alert-danger">
Грубая ошибка, необходимо исправить
</div>


Заранее прошу не удалять мои комментарии, это важно. Если у тебя возникнут вопросы, обозначь их, я с радостью тебе помогу!)

Как вариант, можешь обозначить их так:
<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ СТУДЕНТА:</b>
</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [55]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
#посмотрим структуру и состав данных
print(data.info())
print(data.head(20))
print()
#проверим очевидные ограничения
print('Max возраст',data['dob_years'].max())
print('Min возраст',data['dob_years'].min())
print(data['dob_years'].value_counts())
print()
print(data['children'].value_counts())
print()
print(data['education'].value_counts())
print()
#посмотрим состав уникальных данных. Определим проблемы
print(data['family_status'].value_counts())
print()
print(data['gender'].value_counts())
print()
print(data['purpose'].value_counts())
print()
print(data['income_type'].value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None
    children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   

### Вывод

В качестве исходных данных имеем 21525 записей.

Анализ типов серий данных дает возможность заметить следующие вопросы и выдвинуть гипотезы 
days_employed - общий трудовой стаж в днях имеет тип float64. Скорее всего это не тот тип, так как стаж в днях это целое число. Возможно это не будет создавать проблемы в будущем, так как столбец не участвует в вопросах исселдования.

Предварительный осмотр данных обозначает проблемы с данными days_employed. Мы видим отрицательные числа, дробные и пропуски данных NaN.

В данных возраста есть неявный пропуск данных - нулевой возраст заемщика.

В колонке количество детей есть отрицательное число -1. Максимальное количество детей 20, т.е.данные скорее всего содержат ошибки. Вопрос в исследовании звучит как "Есть ли зависимость между наличием детей и возвратом кредита в срок?", вероятно это не будет иметь значения.

Серия education — уровень образования клиента содержит одни и те же категории данных в разном регистре.

Серия total_income — ежемесячный доход содержит пропущенные данные NaN, хотя они нужны в данном исследовании.

Серия purpose — цель получения кредита содержит одни и те же категории, сформулированные разным языком, что непосредственно в таком виде мешает категоризации. Нужна лемматизация данных.

Пол gender содержит XNA - 1 запись.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
    
Молодец, хороший обзор

</div>

### Шаг 2. Предобработка данных

### Обработка пропусков

In [56]:
#посчитаем записи в которых в значимых колонках пусто
print('Пропуск данных')
print('Дети NA', data['children'].isna().sum())
print('Семейное положение NA', data['family_status'].isna().sum())
print('Уровень дохода NA', data['total_income'].isna().sum())
print('Цели', data['purpose'].isna().sum())
print('Уровень дохода пропущен, но есть долги', len(data[data['total_income'].isna() & data['debt'] == 1]))
print('Уровень дохода пропущен, но есть долги', data[data['total_income'].isna() & data['debt'] == 1])
print('Уровень дохода 0, но есть долги', data[(data['total_income'] == 0) & (data['debt'] == 1)])
print()
#заменим пустой доход на медиану
#при этом посчитаем медиану по каждому виду занятости отдельно
# сформируем сводную табличку с группировкой по типу занятости и измерением медианного дохода
group_income = data.groupby('income_type')['total_income'].median()
# заполним пропуски на медианный доход, основываясь на типе занятости
for i in group_income.index:
    data.loc[(data['income_type'] == i) & (data['total_income'].isnull()), 'total_income'] = group_income[i]
    
#проверим замену
print('Уровень дохода NA', data['total_income'].isna().sum())
print()
#пропуск пола в одной строке проанализируем подробнее
print(data[data['gender'] == 'XNA'])
print()
#удалим его
data = data.drop(data[data['gender'] == 'XNA'].index)
#проверим
print(data['gender'].value_counts())
#все ок

# заполним пропуски возраста на медианный возраст, основываясь на типе занятости 
#Так например можно более точно восстановить возраст пенсионеров, например.
group_dop_years = data.groupby('income_type')['dob_years'].median()
# заполним пропуски на медианный возраст, основываясь на типе занятости
for i in group_dop_years.index:
    data.loc[(data['income_type'] == i) & (data['dob_years'] == 0), 'dob_years'] = group_dop_years[i]

#обработаем стаж
group_days_employed_income = data.groupby('income_type')['days_employed'].mean()
print(group_days_employed_income)
#думаю можно слишком большой стаж поделить на 24, потом на 365, так как похоже на то, что он попросту измерялся в часах. Для 
#адекватных типов занятости мы просто делим на 365. Таким образом, создается новый признак стажа в годах.
#а также избавимся от отрицательных значений при помощи функции abs
data.loc[data['income_type'].str.lower().str.contains('безработный|пенсионер'), 'days_employed'] = abs(data['days_employed'] / 365 /24)
data.loc[data['income_type'].str.lower().str.contains('безработный|пенсионер') == False, 'days_employed'] = abs(data['days_employed'] / 365)

group_days_employed_income = data.groupby('income_type')['days_employed'].mean()
print(group_days_employed_income)

# заполним пропуски days_employed на медианное значение, основываясь на типе занятости
group_days_employed = data.groupby('income_type')['days_employed'].median()
# заполним пропуски на медианный возраст, основываясь на типе занятости
for i in group_days_employed.index:
    data.loc[(data['income_type'] == i) & (data['days_employed'].isnull()), 'days_employed'] = group_days_employed[i]

Пропуск данных
Дети NA 0
Семейное положение NA 0
Уровень дохода NA 2174
Цели 0
Уровень дохода пропущен, но есть долги 170
Уровень дохода пропущен, но есть долги        children  days_employed  dob_years education  education_id  \
55            0            NaN         54   среднее             1   
181           0            NaN         26   среднее             1   
247           1            NaN         60    высшее             0   
278           1            NaN         23   Среднее             1   
312           1            NaN         33   среднее             1   
...         ...            ...        ...       ...           ...   
20592         3            NaN         35   среднее             1   
20646         1            NaN         50   среднее             1   
20917         0            NaN         50    высшее             0   
21271         2            NaN         42   среднее             1   
21426         0            NaN         49   среднее             1   

          

### Вывод

Проверим значимые для нашей задачи поля: Дети, Семейное положение, Уровень дохода, Цели

Пустые значения есть только в уровне дохода

Вероятно 2174 человека просто не указали уровень дохода, хотя не ясно как они при этом прошли скоринг и получили кредит. Возможно тогда это техническая ошбика выгрузки. 10% данных слишком значимо, чтобы такие данные удалять. Проверим сколько таких с долгами - 170 строк, 0.8% Это уже не так много

Проверим, есть ли такие, у кого доход 0, а есть долг. -> Таких нет

Заменим Nan в поле доход на медиану исходя из медианы соотвествующей группы занятости для возможности дальнейшей обработки

Мы имеем одну строку с искажением в поле данных Пол. Одна строка на 21525 несущетсвенна - можно просто удалить. Удалим строку с пустым полом. 

В данных возраста есть неявный пропуск данных - нулевой возраст заемщика. Заменим его на медиану по группам income_type. Так например можно более точно восстановить возраст пенсионеров, например.

Стаж приведен к единому соотвествию. Получен как предполагается стаж в годах. Избавились от отрицательных значений. Затем пропуски заполнены медианным значением, используя группировку по типу занятости

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 

Касаемо пропусков можно было использовать чуть более глубокий подход — группировку!

А именно: от чего у нас зависит стаж \ доход? Как вариант от типа занятости, верно? Или от возрастной группы.
    
Можно было провести группировку по типу занятости с измерением медианного стажа \ дохода, далее заполнить пропуски соответствующими значениями, согласись, так было бы объективнее?)
    
Примерчик:
    
    
    # СПОСОБ 1, ЦИКЛ
    # сформируем сводную табличку с группировкой по типу занятости и измерением медианного дохода

    group_income = df.groupby('income_type')['total_income'].median()


    # заполним пропуски на медианный доход, основываясь на типе занятости
    for i in group_income.index:

        df.loc[(data['income_type'] == i) & (df['total_income'].isnull()), 'total_income'] = group_income[i]
        
        
    # СПОСОБ 2, transform
    #создаем переменные по медиане зарплаты
    med_salary = df.groupby('income_type')['total_income'].transform('median')

    #заменяем пропуски медианой по каждой группе
    df['total_income'] = df['total_income'].fillna(med_salary)
    
</div>
<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ СТУДЕНТА:</b>
    
    Здорово. Спасибо за примеры
    
    единственно так и не смог понять почему работает предложенный во втором варианте код
    а не что-то типа
    
    df['med_salary'] = df.groupby('income_type')['total_income'].transform('median')
    #заменяем пропуски медианой по каждой группе
    df['total_income'] = df['total_income'].fillna(df['med_salary'])
    
    Как происходит параллельная адресаци в df и med_salary в случае
    df['total_income'] = df['total_income'].fillna(med_salary) ??
</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
    
Там происходит соотношение по типу занятости, грубо говоря проверяется есть пропуск \ нет, далее смотрит тип занятости и в зав-ти от этого подставляет нужное значение из таблички med_salary. 

</div>

<div class="alert alert-block alert-danger">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
Твое заполнение пропусков по доходу допустимо (молодец кстати, что использовал медиану, она более устойчива к выбросам), также жирный плюс тебе, что обработал пол. 
    
Однако все же кейс посвящен предобработке, поэтому:
    
* нужно доработать пропуски по стажу (однако счала было бы хорошо обработать аномалии, см.ниже) + замени нулевой возраст (это неявный пропуск)
</div>

<div class="alert alert-block alert-warning">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 

Касаемо признака стажа, ты конечно можешь забить его той же медианой и сказать, что он не используется, так что можно проигнорировать его, но мне бы хотелось, чтобы ты с ним поработал более подробно, тк это прокачивает навыки:
    
* если ты используешь подсчет среднего стажа по типам занятости, то заметишь, что у нас есть аномалии положительного и отрицательного стажа. Что примечательно положительный стаж только у некоторых типов занятости + он же выглядит слишком большим.
    
* думаю можно слишком большой стаж поделить на 24, потом на 365, так как похоже на то, что он попросту измерялся в часах. Для адекватных типов занятости мы просто делим на 365. Таким образом, создается новый признак стажа в годах.
    
* после наша задача избавиться от отрицательных значений, поможет abs
    
* и на последок необходимо заполнить пропуски, можно использовать ту же группировку по типу занятости с измерением медианы
</div>

<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ СТУДЕНТА:</b>

- думаю можно слишком большой стаж поделить на 24, потом на 365, так как похоже на то, что он попросту измерялся в часах. Для 
- адекватных типов занятости мы просто делим на 365. Таким образом, создается новый признак стажа в годах.
    
Вот здесь не пойму следующее. Сама логика и преобразования ясны. Но ведь это все лишь умозаключения. Разве "научно точно" применять разные преобразования к одной колонке, основываясь на личных умозаключениях? Так можно ошибиться в логике и сильно разбалансировать данные. Разве нет?
    
Еще если поделить стаж предпринимателя средний на 365 мы получим 1,4 года. По сути и госслужащий и компаньон тоже получают малый стаж в годах. Кажется логика не точна?!

</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
    
Это безусловно один из возможных сценариев обработки, я не говорил, что он верный, но в любом случае он приводит стаж к чему-то более приближенному к жизни + прокачивает твои навыки. В реальной жизни безусловно мы бы пошли к ребятам, которые занимались базой данных, и проконсультировались. Касаемо стажа год+, это вполне могут быть клиенты, которые работают мало по своему типу занятости. Признак стажа на текущем исследовании не нужен, моей целью было приведение его хотя бы чуточку в более менее приемлемый вид + прокачка твоих навыков, ведь кейс посвящен предобработке


</div>

### Замена типа данных

In [57]:
#мы определяли проблему в типе даннных days_employed
#заменим дробный тип на целый
#data['days_employed'] = data['days_employed'].astype('int64')

data['total_income'] = data['total_income'].astype('int')
data['days_employed'] = data['days_employed'].astype('int')
print(data.info())
print(data.head(20))

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
children            21524 non-null int64
days_employed       21524 non-null int64
dob_years           21524 non-null float64
education           21524 non-null object
education_id        21524 non-null int64
family_status       21524 non-null object
family_status_id    21524 non-null int64
gender              21524 non-null object
income_type         21524 non-null object
debt                21524 non-null int64
total_income        21524 non-null int64
purpose             21524 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.8+ MB
None
    children  days_employed  dob_years            education  education_id  \
0          1             23       42.0               высшее             0   
1          1             11       36.0              среднее             1   
2          0             15       33.0              Среднее             1   
3          3     

### Вывод

Колонки дохода и дни стажа преобразованы к целому типу.
Изначально был тип с плавающей точкой. Это увеличивает обьем памяти, а копейки в доходе - несущественная информация для исследования

<div class="alert alert-block alert-danger">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
Все же тип стажа нужно сконвертировать, понимаешь, ты должен научиться всячески восстанавливать данные, это полезный навык
    
</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
    
Возраст тоже стоило бы привести к типу int, но не критично

</div>

### Обработка дубликатов

In [58]:
#дубликаты при первичном осмотре обнаружены в столбцах
#'education' и 'purpose'
#для этого использовался метод value_counts()
#Посмотрим кол-во полных дубликатов ПЕРЕД приведением регистра, после перепроверь их кол-во ПОСЛЕ обработки регистра
print('Количество полных дубликатов', data.duplicated().sum())
#проведем приведение к нижнему регистру столбца 'education', так как дубликаты в нем существуют из-за разного регистра слов
data['education'] = data['education'].str.lower()
print()
print(data['education'].value_counts())
print()
#проверим остались ли дубликаты после преобразования
print(data['purpose'].value_counts())
print()
#
data['children'] = data['children'].replace(-1,1)
#20 возникло из-за лишнего нуля или нет
#поэтому заменим его на медиану
data['children'] = data['children'].replace(20,data['children'].median())
print(data['children'].value_counts())
print()
#Посмотрим кол-во полных дубликатов ПОСЛЕ обработки регистра
print('Количество полных дубликатов', data.duplicated().sum())
#ничего не поменялось удалим их
data = data.drop_duplicates().reset_index(drop=True)

Количество полных дубликатов 54

среднее                15233
высшее                  5260
неоконченное высшее      743
начальное                282
ученая степень             6
Name: education, dtype: int64

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      623
покупка своего жилья                      620
строитель

### Вывод

Из 15 вариантов написания получили всего 5. Теперь дубликатов в колонке education нет.

Цели кредита все уже в нижнем регистре. Для обработки дубликатов нужна лемматизация на следующем шаге

В данных по детям устранены некорректные значения (-1,20)

Удалены полные дубликаты строк

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
Молодец, что привел признак образования к нижнему регистру, а также обработал аномалии по детям, однако ознакомься с замечанием ниже
    
</div>
<div class="alert alert-block alert-danger">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
    
* Посмотри кол-во полных дубликатов `df.deplucated().sum()` ПЕРЕД приведением регистра, после перепроверь их кол-во ПОСЛЕ обработки регистра
    
* Удали дублирующиеся строки
</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
    
Хорошо, молодец, как видим единообразие регистра приводит к детектированию дополнительных дублей

</div>

### Лемматизация

In [59]:
#определим категории в процессе лематизации
from pymystem3 import Mystem
import collections
m = Mystem()

data_purposes = data['purpose'].unique()
lemmas_count = collections.Counter()

for purpose in data_purposes:
    lemmatize_purpose = m.lemmatize(purpose)
    print(lemmatize_purpose)
    for word in lemmatize_purpose:
        lemmas_count[word] += 1

print()
print(lemmas_count)

['покупка', ' ', 'жилье', '\n']
['приобретение', ' ', 'автомобиль', '\n']
['дополнительный', ' ', 'образование', '\n']
['сыграть', ' ', 'свадьба', '\n']
['операция', ' ', 'с', ' ', 'жилье', '\n']
['образование', '\n']
['на', ' ', 'проведение', ' ', 'свадьба', '\n']
['покупка', ' ', 'жилье', ' ', 'для', ' ', 'семья', '\n']
['покупка', ' ', 'недвижимость', '\n']
['покупка', ' ', 'коммерческий', ' ', 'недвижимость', '\n']
['покупка', ' ', 'жилой', ' ', 'недвижимость', '\n']
['строительство', ' ', 'собственный', ' ', 'недвижимость', '\n']
['недвижимость', '\n']
['строительство', ' ', 'недвижимость', '\n']
['на', ' ', 'покупка', ' ', 'подержать', ' ', 'автомобиль', '\n']
['на', ' ', 'покупка', ' ', 'свой', ' ', 'автомобиль', '\n']
['операция', ' ', 'с', ' ', 'коммерческий', ' ', 'недвижимость', '\n']
['строительство', ' ', 'жилой', ' ', 'недвижимость', '\n']
['жилье', '\n']
['операция', ' ', 'со', ' ', 'свой', ' ', 'недвижимость', '\n']
['автомобиль', '\n']
['заниматься', ' ', 'образование'

### Вывод

Мы видим что в категориях должы участвовать слова 'покупка': 10, 'недвижимость': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, которые встречаются максимальное число раз. Это подтверждается визуально при оценке уникальных значений из списка целей кредита

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
    
Молодец, хорошо выделил наиболее популярные леммы, думаю можно категоризировать!

</div>

### Категоризация данных

In [60]:
#нам нужна функция чтобы определить категории в таблице
def purpose_category(purpose):
    purpose_l = m.lemmatize(purpose)
    if ('недвижимость' in purpose_l) or ('жилье' in purpose_l):
        prefix = 'приобретение '
        #if 'ремонт' in purpose_l:
        #    prefix = 'ремонт '
        #elif 'строительство' in purpose_l:
        #    prefix = 'строительство '
        #else:
        #    prefix = 'приобретение '
        #
        #if 'жилой' in purpose_l:
        #    return prefix+'жилой недвижимости'
        #elif 'свой' in purpose_l:
        #    return prefix+'жилой недвижимости'
        #elif 'коммерческий' in purpose_l:
        #    return prefix+'коммерческой недвижимости'
        #elif 'сдача' in purpose_l:
        #    return prefix+'коммерческой недвижимости'
        #else:
        return prefix+'недвижимости'
    elif 'автомобиль' in purpose_l:
        return 'автомобиль'
    elif 'образование' in purpose_l:
        return 'образование'
    elif 'свадьба' in purpose_l:
        return 'свадьба'
    
def income_category(income:int, p25:int, p50:int, p75:int):
    if (income <= p25):
        return 'низкий доход'
    elif (p25 < income) and (income <= p50):
        return 'средний доход'
    elif (p50 < income) and (income <= p75):
        return 'выше среднего'
    else:
        return 'высокий доход'

#добавим столбец с категоирей цели
data['purpose_category'] = data['purpose'].apply(purpose_category)
#
print(data['purpose_category'].value_counts())
#Вывести статистики с помощью describe
data_descr = data.describe()
data_descr['total_income'] = data_descr['total_income'].map(lambda x: '{:.0f}'.format(x))
display(data_descr)
#print(data_descr.info())
perc25 = int(data_descr.loc['25%', 'total_income'])
perc50 = int(data_descr.loc['50%', 'total_income'])
perc75 = int(data_descr.loc['75%', 'total_income'])
#категоризируем уровень дохода в группы по перцептилям
#display(data.info())
data['total_income_category'] = data['total_income'].apply(income_category, args=(perc25, perc50, perc75))
print()
print(data['total_income_category'].value_counts())

приобретение недвижимости    10810
автомобиль                    4306
образование                   4013
свадьба                       2324
Name: purpose_category, dtype: int64


children  days_employed     dob_years  education_id  \
count  21453.000000   21453.000000  21453.000000  21453.000000   
mean       0.473500      12.094625     43.475598      0.817042   
std        0.751162      14.636713     12.226444      0.548628   
min        0.000000       0.000000     19.000000      0.000000   
25%        0.000000       2.000000     33.000000      1.000000   
50%        0.000000       5.000000     42.000000      1.000000   
75%        1.000000      14.000000     53.000000      1.000000   
max        5.000000      50.000000     75.000000      4.000000   

       family_status_id          debt total_income  
count      21453.000000  21453.000000        21453  
mean           0.973896      0.081154       165318  
std            1.421601      0.273078        98189  
min            0.000000      0.000000        20667  
25%            0.000000      0.000000       107620  
50%            0.000000      0.000000       142594  
75%            1.000000      0.000000       195818  
max            4.000000      1.000000      2265604


средний доход    5479
низкий доход     5364
высокий доход    5363
выше среднего    5247
Name: total_income_category, dtype: int64


### Вывод

Данные категоризированы по цели кредита с использованием словаря следующих категорий:
приобретение недвижимости                 
автомобиль                                
образование                               
свадьба                                                  

для этого использовано применение функции purpose_category() к колонке purpose

Данные категоризированы по доходу заемщика на категории кратные 100 000 тыс руб. При этом стало ясно что выше 10 категории дохода не имеет смысла рассматривать отдельно. Обьеденим их при помощи функции

Рассмотрев статистику с помощью describe, видно, что распределение данных по доходам очень хорошо делится квартилями. Иными словами, все что до 1 квартиля (25%) — низкий доход, между 1 и мединой(2 квартиль) тоже будет 25% — средний доход, от медианы до 3 квартиля (25%) — выше среднего, а все что дальше 3 квартиля — высокий доход.
Категоризируем данные согласно этой логике.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
    
В целом хорошая категоризация по целям кредита, но возможно некоторые категории можно было объединить в одну (строительство  + ремонт)

</div>

<div class="alert alert-block alert-danger">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
Думаю, что категории по доходу стоит сделать более равномерными, тем самым более репрезентативными:
    
1. Вывести статистики с помощью describe, так бы сразу все было видно

2. Либо построить ящик с усами https://plotly.com/python/box-plots/
    
3. У нас квартилями распределение очень хорошо делится на 25%. Иными словами, все что до 1 квартиля (25%) — низкий доход, между 1 и мединой(2 квартиль) тоже будет 25% — средний доход, от медианы до 3 квартиля (25%) — выше среднего, а все что дальше 3 квартиля — высокий доход. https://habr.com/ru/post/267123/

</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
    
Молодец, хорошая доработка!

</div>

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [61]:
debt_children = pd.DataFrame()
debt_children['debt_children_count'] = data.groupby('children')['debt'].count()
debt_children['debt_children_sum'] = data.groupby('children')['debt'].sum()
debt_children['debt_children_ratio'] = (debt_children['debt_children_sum']/debt_children['debt_children_count']).map(lambda x: '{:.1%}'.format(x))
debt_children = debt_children.sort_values(by='debt_children_ratio', ascending=False)
#
debt_children

debt_children_count  debt_children_sum debt_children_ratio
children                                                            
4                          41                  4                9.8%
2                        2052                194                9.5%
1                        4855                445                9.2%
3                         330                 27                8.2%
0                       14166               1071                7.6%
5                           9                  0                0.0%

### Вывод

С одной стороны количество просроченных кредитов растет с ростом количества детей. С другой стороны заемщики с 3 детьми платят лучше, чем с 1. В целом уровень просроченных кредитов у заемщиком с детьми выше чем без детей

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 

* Как мы видим, у нас межу 0 и 1 ребенком происходит довольно ощутимый скачок невозвратов, с 7 до 9 %, делаем вывод, что наличие детей, именно наличие, играет существенную роль. А вот если мы начнем изучать кол-во детей, то заметим, что % варьируется от ребенка к ребенку, но он не такой существенный как факт наличия ребенка, но все же также значим. Видим, что выделяется группа семей с тремя детьми, она наиболее безопасная (если не учитывать семьи без детей)
    
P.s. к слову можно было провести категоризацию по двум направлениям: есть дети \ нет детей; бездетная \ малодетная \ многодетная
    
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [62]:
debt_family_status = pd.DataFrame()
debt_family_status['debt_family_status_count'] = data.groupby('family_status')['debt'].count()
debt_family_status['debt_family_status_sum'] = data.groupby('family_status')['debt'].sum()
debt_family_status['debt_family_status_ratio'] = (debt_family_status['debt_family_status_sum']/debt_family_status['debt_family_status_count']).map(lambda x: '{:.1%}'.format(x))
debt_family_status = debt_family_status.sort_values(by='debt_family_status_ratio', ascending=False)
#
debt_family_status

debt_family_status_count  debt_family_status_sum  \
family_status                                                             
Не женат / не замужем                      2810                     274   
гражданский брак                           4150                     388   
женат / замужем                           12339                     931   
в разводе                                  1195                      85   
вдовец / вдова                              959                      63   

                      debt_family_status_ratio  
family_status                                   
Не женат / не замужем                     9.8%  
гражданский брак                          9.3%  
женат / замужем                           7.5%  
в разводе                                 7.1%  
вдовец / вдова                            6.6%

### Вывод

Одинокие заемщики (в разводе - 7.1%, вдовец / вдова - 6.6%) или заемщики в оформленных отношениях (женат / замужем - 7.5%) имеют просрочку реже, чем не Не женат / не замужем - 9.7% или гражданский брак - 9.3%

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 

* Как можно заметить, статус официальности отношений действительно положительно сказывается на возврате кредита. Люди, которые не состоят в официальных отношениях и никогда не состояли хуже возвращают кредиты, в отличии от семей, которые имеют официальный статус, пусть даже когда-то в прошлом.
    
P.s. ты знал, что можно чуть более эффективно строить сводные таблички? Смотри:
    
</div>

In [63]:
# КОД РЕВЬЮВЕРА

import plotly.express as px

def pivot(data, index):
    
    """
    Я решил написать функцию, которая строит сводную табличку и визуализирует ее, тут нет ничего сложного, смотри:
    на вход она принимает датафрейм и признак для группировки, далее строит сводную с подсчетом клиентов в категории (count),
    числа заемщиков (sum), % невозврата (mean). Тк признак debt бинарный, то такие аггрегирующие функции как нельзя кстати.
    
    
    count - подсчитает чисто все строки, т.е.  общее число клиентов
    sum - сумму ноликов и единичек, т.е. по факту будет только сумма единиц => кол-во должников
    mean - сумма деленна на кол-во, т.е. те же должники деленные на общее число клиентов
    
    Классно же!)
    
    """
    
    pivot = data.pivot_table(index = index, values = 'debt', aggfunc = ['count', 'sum', 'mean']).reset_index()
    pivot.columns = [index, 'Кол-во клиентов', 'Кол-во должников', '% невозврата']
    display(pivot.sort_values('% невозврата', ascending = False))
    
    fig = px.bar(pivot, x = index, y = '% невозврата', color = '% невозврата', title = '% невозврата по ' + index)
    fig.show()
    

pivot(data, 'family_status')

family_status  Кол-во клиентов  Кол-во должников  % невозврата
0  Не женат / не замужем             2810               274      0.097509
3       гражданский брак             4150               388      0.093494
4        женат / замужем            12339               931      0.075452
1              в разводе             1195                85      0.071130
2         вдовец / вдова              959                63      0.065693

<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ СТУДЕНТА:</b>
    
Действительно лаконичный код получается. За это отдельное спс
    
Интересно у меня в какой-то момент перестала отображаться гистограмма. И кэши чистил и сайт обновлял. Нет. В чем обычно причина?
    
</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
    
Это особенность работы библиотеки, хотя я бы даже сказал недостаток, то остается, то исчезает. В любом случае, это лечится обычным перезапуском ячейки

</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [64]:
debt_income = pd.DataFrame()
debt_income['total_income_category_count'] = data.groupby('total_income_category')['debt'].count()
debt_income['total_income_category_sum'] = data.groupby('total_income_category')['debt'].sum()
debt_income['total_income_category_ratio'] = (debt_income['total_income_category_sum']/debt_income['total_income_category_count']).map(lambda x: '{:.1%}'.format(x))
debt_income = debt_income.sort_values(by='total_income_category_ratio', ascending=False)
#
debt_income

total_income_category_count  total_income_category_sum  \
total_income_category                                                           
средний доход                                 5479                        483   
выше среднего                                 5247                        448   
низкий доход                                  5364                        427   
высокий доход                                 5363                        383   

                      total_income_category_ratio  
total_income_category                              
средний доход                                8.8%  
выше среднего                                8.5%  
низкий доход                                 8.0%  
высокий доход                                7.1%

### Вывод

Интересно, но люди со средним и выше среднего уровнем дохода чаще оказываются в должниках, чем с низким или высоким.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
    
Да, это действительно так, сам удивлен)

</div>

<div class="alert alert-block alert-danger">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
Воспользуйся моим советом и пересмотри ответ на вопрос с учетом корректировки категорий
    
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [65]:
debt_purposes= pd.DataFrame()
debt_purposes['debt_purposes_count'] = data.groupby('purpose_category')['debt'].count()
debt_purposes['debt_purposes_sum'] = data.groupby('purpose_category')['debt'].sum()
debt_purposes['debt_purposes_ratio'] = (debt_purposes['debt_purposes_sum']/debt_purposes['debt_purposes_count']).map(lambda x: '{:.1%}'.format(x))
debt_purposes = debt_purposes.sort_values(by='debt_purposes_ratio', ascending=False)
#
debt_purposes

debt_purposes_count  debt_purposes_sum  \
purpose_category                                                    
автомобиль                                4306                403   
образование                               4013                370   
свадьба                                   2324                186   
приобретение недвижимости                10810                782   

                          debt_purposes_ratio  
purpose_category                               
автомобиль                               9.4%  
образование                              9.2%  
свадьба                                  8.0%  
приобретение недвижимости                7.2%

### Вывод

Из таблицы мы видим зависимость процента невозврата кредита в зависимости от целей его получения. Кредиты на покупку авто и образование имеют максимальный % невозврата - 9.3% и 9.2% соотвественно. Приобретение недвижимости и ремонт наименьшее - 7.2% и 5.7% соответсвенно.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 

* Согласен, видим, что наиболее рисковые цели это у нас образование и автомобиль, тогда как свадьба и недвижимость напротив, являются наиболее безопасными целями

</div>

### Шаг 4. Общий вывод

Была рассмотрена зависимость невозврата кредита от наличия детей у заемщика, от семейного положения, уровня дохода и цели на которую данный кредит брался.

Обнаружена заисимость невозврата от наличия детей. В целом уровень просроченных кредитов у заемщиком с детьми выше чем без детей. 

Существует зависимость невозврата кредита от семейного положения. Одинокие заемщики или заемщики в оформленных отношениях имеют просрочку реже, чем не женатые или в гражданском браке.

Кредиты на покупку авто и образование имеют максимальный % невозврата. Приобретение недвижимости наименьшее.

Люди со средним и выше среднего уровнем дохода чаще оказываются в должниках, чем с низким или высоким.

<div class="alert alert-block alert-warning">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 🎄:</b> 
    
`Родион, ты хорошо выполнил проект, но подправь моменты, на которые я указал, твои выводы местами чуть изменятся`
    
</div>

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.